### PR TITLE
[TT-12587] verify root CA on certificate check mw

### PR DESCRIPTION
**Pull request by jeffy-mathew**
## Description
Verify uploaded root CA certificate against client certificate in certificate check middleware.
Loop through peer certificate and do the checks so that intermediate CAs can also be added to allow list.

## Related Issue
https://tyktech.atlassian.net/browse/TT-12587

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [ ] I ensured that the documentation is up to date
- [ ] I explained why this PR updates go.mod in detail with reasoning why it's required
- [ ] I would like a code coverage CI quality gate exception and have explained why

___

### **PR Type**
Enhancement, Bug fix

___

### **Description**
- Added functionality to verify the uploaded root CA certificate against the client certificate in the certificate check middleware.
- Refactored the existing logic to simplify peer certificate handling.
- Introduced a new helper function `verifyCertAgainstCA` to encapsulate the CA verification logic.
- Improved error handling for scenarios where certificates are invalid or expired.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>helpers.go</strong><dd><code>Enhance mTLS validation with CA certificate verification</code>&nbsp; </dd></summary>
<hr>

internal/crypto/helpers.go

<li>Added verification of client certificate against CA certificate.<br> <li> Refactored the logic to handle peer certificates.<br> <li> Introduced <code>verifyCertAgainstCA</code> function to encapsulate CA <br>verification.<br> <li> Improved error handling for certificate validation.<br>

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6422/files#diff-3d1fc755c46eaa99e9f1edf358b9e00842342ae6333902959d7a68b46d156829">+29/-22</a>&nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: internal/crypto/helpers.go [hunk 1]
```diff
@@ -93,32 +93,39 @@ func ValidateRequestCerts(r *http.Request, certs []*tls.Certificate) error {
...
+
+		if cert.Leaf.IsCA && verifyCertAgainstCA(cert, peerCertificate) {
+			return nil
+		}
+	}
+
+	return errors.New("Certificate with SHA256 " + certID + " not allowed")
+}
+
+func verifyCertAgainstCA(caCert *tls.Certificate, peerCert *x509.Certificate) bool {
```
Comment (titpetric) [inline internal/crypto/helpers.go:116]: move to internal/crypto?

File: internal/crypto/helpers.go [hunk 2]
```diff
@@ -93,32 +93,39 @@ func ValidateRequestCerts(r *http.Request, certs []*tls.Certificate) error {
...
+
+	// Create a verification options struct
+	opts := x509.VerifyOptions{
+		Roots: roots,
 	}
 
-	return errors.New("Certificate with SHA256 " + HexSHA256(r.TLS.PeerCertificates[0].Raw) + " not allowed")
+	// Verify the client certificate
+	_, err := peerCert.Verify(opts)
+	return err == nil
```
Comment (titpetric) [inline internal/crypto/helpers.go:128]:
```suggestion
	return err
```

File: internal/crypto/helpers.go (hunk 1 above)
Comment (titpetric) [inline internal/crypto/helpers.go:116]:
```suggestion
func verifyCertAgainstCA(caCert *tls.Certificate, peerCert *x509.Certificate) error {
```

Comment (jeffy-mathew) [reply internal/crypto/helpers.go:116]: it's already in `internal/crypto/helpers.go`. do you mean something else?

Comment (titpetric) [reply internal/crypto/helpers.go:125]: misread it

**Review by titpetric — APPROVED** [review]

**Review by titpetric — APPROVED** [review]

**Review by titpetric — APPROVED** [review]
Changes LGTM, add a code comment for the use case around L100 so we know to consider it in the future;

**Pull request merged**